vcf.c

```diff
@@ -4448,8 +4448,10 @@ int bcf_update_alleles(const bcf_hdr_t *hdr, bcf1_t *line, const char **alleles,
 
     // If the supplied alleles are not pointers to line->d.als, the existing block can be reused.
     int i;
+    uintptr_t als_start = (uintptr_t) line->d.als;
+    uintptr_t als_end = line->d.als ? als_start + line->d.m_als : als_start;
```

Comment:
So just to clarify what I was trying to say earlier but obviously not explaining it well, the above ?: is pointless and just obfuscates the code.

I get what it's trying to achieve: adding offset to pointer is invalid is pointer is NULL, even if offset is zero.  However you've also converted these to uintptr_t already.  `als_start` isn't a pointer, it's an integer.  Adding 0 to 0 is 100% legal and defined behaviour, so we don't need to first check whether the value is zero with a ?: operator.  It could just be `als_end = als_start + line->d.m_als`.

Given that, it's clear to see that both local variables are also unnecessary and could simply be replaced by a cast in the if statement as that's now all they're doing.

Reply:
This has now been replaced with a more comprehensive solution that completely eliminates the undefined behaviour.  One advantage with the new version is that it works better with the way bcftools calls the function (e.g. in [norm](https://github.com/samtools/bcftools/blob/b626e984/vcfnorm.c#L162)) and makes it much more likely to reuse the existing buffer.  When I tested, it halved the amount of time taken in `bcf_update_alleles()` (although it doesn't affect the run time too much as most time is spent elsewhere).